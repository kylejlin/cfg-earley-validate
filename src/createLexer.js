const createLexer = (tokenTypes) => {
  tokenTypes = tokenTypes.map((type) => {
    const { name, rule, value, isOmitted } = type;
    const caretRule = new RegExp('^' + rule.source);
    if (caretRule.test('')) {
      throw new TypeError('Rule ' + rule.source + ' for token type ' + name + ' matched the empty string. Rules cannot match the empty string.');
    }
    return {
      name,
      rule: caretRule,
      value: 'function' === typeof value ? value : null,
      isOmitted,
    };
  });

  return (string) => {
    const tokens = [];
    let line = 1;
    let column = 0;
    while (string.length > 0) {
      let hasFoundRule = false;

      for (const type of tokenTypes) {
        const { name, rule, value: valueMapper, isOmitted } = type;
        const match = string.match(rule);
        if (match === null) {
          continue;
        }
        const [firstMatch] = match;

        string = string.slice(firstMatch.length);

        const startLine = line;
        const startColumn = column;
        const newlines = firstMatch.split('\n').length - 1;
        if (newlines > 0) {
          line += newlines;
          column = firstMatch.split('\n').slice(-1)[0].length;
        } else {
          column += firstMatch.length;
        }

        if (!isOmitted) {
          const value = valueMapper === null
            ? firstMatch
            : valueMapper(firstMatch);
          tokens.push({
            name: name,
            value,
            start: [startLine, startColumn],
            end: [line, column],
          });
        }

        hasFoundRule = true;
        break;
      }

      if (!hasFoundRule) {
        throw new SyntaxError('Cannot tokenize: ' + string);
      }
    }
    return tokens;
  };
};

export default createLexer;
