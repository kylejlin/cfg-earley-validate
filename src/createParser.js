const createParser = (grammar) => {
  grammar = grammar.map(parseRule);

  return (tokens) => {
    tokens = tokens.concat([{ name: 'END_OF_INPUT' }]);
    const chart = Array(tokens.length).fill(null).map(() => []);
    const startRule = grammar[0];
    const startState = [startRule[0], [], startRule.slice(1), 0];
    chart[0].push(startState);
    for (let i = 0; i < tokens.length; i++) {
      while (true) {
        let changes = false;
        for (const state of chart[i]) {
          const [x, ab, cd, j] = state;
          const nextStates0 = getClosure(grammar, i, x, ab, cd, j);
          for (const nextState of nextStates0) {
            changes = changes || addToChart(chart, i, nextState);
          }
          const nextState = getShift(tokens, i, x, ab, cd, j);
          if (nextState !== null) {
            changes = changes || addToChart(chart, i + 1, nextState);
          }
          const nextStates1 = getReductions(chart, i, x, ab, cd, j);
          for (const nextState of nextStates1) {
            changes = changes || addToChart(chart, i, nextState);
          }
        }
        if (!changes) {
          break;
        }
      }
    }
    const acceptingState = [startRule[0], startRule.slice(1), [], 0];
    return !!chart[tokens.length - 1].find((s) => deepEq(s, acceptingState));
  };
};

const parseRule = (rule) => rule.split(/ |:/).filter((s) => s !== '');

const deepEq = (a, b) => {
  if (Array.isArray(a)) {
    if (!Array.isArray(b)) {
      return false;
    }
    return a.length === b.length && a.every((x, i) => deepEq(x, b[i]));
  }
  return a === b;
};

const addToChart = (chart, index, elem) => {
  if (!chart[index].find((a) => deepEq(a, elem))) {
    chart[index].push(elem);
    return true;
  }
  return false;
};

const getClosure = (grammar, i, x, ab, cd) => grammar
  .filter((rule) => rule[0] === cd[0])
  .map((rule) => [rule[0], [], rule.slice(1), i]);

const getShift = (tokens, i, x, ab, cd, j) => {
  if (cd.length === 0 || cd[0] !== tokens[i].name) {
    return null;
  }
  return [x, ab.concat(cd.slice(0, 1)), cd.slice(1), j];
};

const getReductions = (chart, i, x, ab, cd, j) => {
  if (cd.length !== 0) {
    return [];
  }
  const reductions = [];
  for (const rule of chart[j]) {
    const [rx, rab, rcd, rj] = rule;
    if (rcd.length === 0 || x !== rcd[0]) {
      continue;
    }
    reductions.push([rx, rab.concat(rcd.slice(0, 1)), rcd.slice(1), rj]);
  }
  return reductions;
};

export default createParser;
