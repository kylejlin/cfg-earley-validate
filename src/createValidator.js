const createValidator = (lexer, parser) => {
  return (code) => {
    try {
      const tokens = lexer(code);
      try {
        return parser(tokens);
      } catch (_) {
        return false;
      }
    } catch (_) {
      return false;
    }
  };
};

export default createValidator;
