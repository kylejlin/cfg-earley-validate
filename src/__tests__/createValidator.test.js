import createLexer from '../createLexer';
import createParser from '../createParser';
import createValidator from '../createValidator';

test('it validates correctly', () => {
  const tokenTypes = [
    {
      name: 'LPAREN',
      rule: /\(/,
    },
    {
      name: 'RPAREN',
      rule: /\)/,
    },
    {
      name: 'NUMBER',
      rule: /\d+(?:\.\d+)?/,
      value: parseFloat,
    },
    {
      name: 'STRING',
      rule: /"(?:[^"\\]|\\.)*"/,
      value: (value) => value.slice(1, -1),
    },
    {
      name: 'IDENTIFIER',
      rule: /[a-zA-Z_][a-zA-Z_0-9]*/,
    },
    {
      isOmitted: true,
      name: 'WHITESPACE',
      rule: /[ \t\n]+/,
    }
  ];
  const lexer = createLexer(tokenTypes);
  const grammar = [
    'start : exp',
    'exp : NUMBER',
    'exp : STRING',
    'exp : IDENTIFIER',
    'exp : funcCall',
    'funcCall : LPAREN IDENTIFIER optArgs RPAREN',
    'optArgs :',
    'optArgs : exp optArgs'
  ];
  const parser = createParser(grammar);
  const validate = createValidator(lexer, parser);
  const code = '(foo 42 (baz "bar"))';
  expect(validate(code)).toMatchSnapshot();
});
