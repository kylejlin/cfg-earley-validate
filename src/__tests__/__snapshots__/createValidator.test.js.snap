// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it validates correctly 1`] = `true`;
